compiler/optimizer/VPHandlers.cpp

```diff
@@ -2653,10 +2653,39 @@ TR::Node *constrainIaload(OMR::ValuePropagation *vp, TR::Node *node)
 #endif
 
 
-  if (node->isNonNull())
+   if (node->isNonNull())
       vp->addBlockConstraint(node, TR::VPNonNullObject::create(vp));
    else if (node->isNull())
       vp->addBlockConstraint(node, TR::VPNullObject::create(vp));
+   else if (node->getSymbolReference())
+      {
+      if (!node->getSymbolReference()->isUnresolved() &&
+          (node->getSymbolReference()->getSymbol()->getKind() == TR::Symbol::IsShadow) &&
+          (node->getSymbolReference()->getCPIndex() >= 0) &&
+          node->getSymbolReference()->getOwningMethod(vp->comp()))
+         {
+         int32_t fieldSigLen;
+         const char *fieldSig = node->getSymbolReference()->getOwningMethod(vp->comp())->fieldSignatureChars(
+                                      node->getSymbolReference()->getCPIndex(), fieldSigLen);
+         int32_t fieldNameLen;
+         const char *fieldName = node->getSymbolReference()->getOwningMethod(vp->comp())->fieldNameChars(
+                                      node->getSymbolReference()->getCPIndex(), fieldNameLen);
+         char *className = node->getSymbolReference()->getOwningMethod(vp->comp())->classNameChars();
+         uint16_t classNameLen = node->getSymbolReference()->getOwningMethod(vp->comp())->classNameLength();
+         if (fieldName && !strncmp(fieldName, "this$0", 6) &&
+              className && !strncmp(className, "java/util/", 10))
```

Comment:
This seems like very Java specific code that should IMO belong in OpenJ9.

Reply:
Moved Java specific code inside  `J9_PROJECT_SPECIFIC` block as @0xdaryl suggested.